<a href="https://colab.research.google.com/github/nabilnehme/AAI_634O/blob/main/Week2/LAU_AAI_634O_A11_202520_Nabil_Nehme_Week2_Schema_Design_and_Indexing_in_MongoDB.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Part1


*   **Should comments be embedded within posts or stored as a separate collection?**

Because the comments may get too big and we sometimes have to perform certain actions on them separately, it is preferable to keep them as a separate collections.

*   **Should tags be referenced or embedded within posts?**

Tags also should be stored as a seperate collection to avoid redundancy and maintain better flexibility.



In [170]:
pip install pymongo

In [171]:
from pymongo import MongoClient

# Replace with your own connection string
connection_string = "mongodb+srv://nabilnehme:nabilnehme@cluster0.51s8b.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0"

# Connect to the MongoDB Atlas cluster
client = MongoClient(connection_string)

# Connect to MongoDB
db = client['blog_platform']

users = db['users']
posts = db['posts']

In [172]:
# CLEAR COLLECTION BEFORE INSERTING
users.delete_many({})  # Deletes all documents in the collection
posts.delete_many({})  # Deletes all documents in the collection
print("Collection cleared successfully!")

Collection cleared successfully!


In [173]:
# Insert sample users
users.insert_many([
    {"name": "Alice", "email": "alice@example.com"},
    {"name": "Bob", "email": "bob@example.com"}
])

InsertManyResult([ObjectId('679f7ec9b68c8ab48b23567e'), ObjectId('679f7ec9b68c8ab48b23567f')], acknowledged=True)

In [174]:
# Insert sample posts with comments and tags

posts.insert_many([
  {
  "title": "How to Use MongoDB",
  "content": "This is a guide to using MongoDB.",
  "author": "Alice",
  "comments": [
    {"user_id": "Bob", "text": "Great post!", "timestamp": "2024-09-12T10:00:00"}
  ],
  "tags": ["MongoDB", "Database"]
  }
])

InsertManyResult([ObjectId('679f7ecab68c8ab48b235680')], acknowledged=True)

In [175]:
# Insert more users
users.insert_many([
    {"name": "Charlie", "email": "charlie@example.com"},
    {"name": "David", "email": "david@example.com"}
])
print("New users inserted successfully!")


New users inserted successfully!


In [176]:
# Adding more comments to specific post
posts.update_one(
    {"title": "How to Use MongoDB"},  # Find the post by title
    {"$push": {"comments":
        {"user_id": "Charlie", "text": "Very helpful, thanks!", "timestamp": "2024-09-13T12:30:00"}
    }}
)

posts.update_one(
    {"title": "How to Use MongoDB"},
    {"$push": {"comments":
        {"user_id": "David", "text": "Nice explanation!", "timestamp": "2024-09-14T15:45:00"}
    }}
)
print("New comments added successfully!")


New comments added successfully!


In [177]:
# Adding more tagss to specific post
posts.update_one(
    {"title": "How to Use MongoDB"},
    {"$addToSet": {"tags": {"$each": ["NoSQL", "Tutorial"]}}}  # Avoids duplicate tags
)

print("New tags added successfully!")


New tags added successfully!


#Part 3: Indexing for Performance

Query Optimization

In [178]:
# Create index on the "author" field
posts.create_index([("author", 1)])
print("Index created on 'author' field!")


Index created on 'author' field!


In [179]:
# Query to fetch posts by author
#author_name = "Bob"
author_name = "Alice"
posts_by_author = posts.find({"author": author_name})


In [180]:
# Print results
for post in posts_by_author:
    print(post)

{'_id': ObjectId('679f7ecab68c8ab48b235680'), 'title': 'How to Use MongoDB', 'content': 'This is a guide to using MongoDB.', 'author': 'Alice', 'comments': [{'user_id': 'Bob', 'text': 'Great post!', 'timestamp': '2024-09-12T10:00:00'}, {'user_id': 'Charlie', 'text': 'Very helpful, thanks!', 'timestamp': '2024-09-13T12:30:00'}, {'user_id': 'David', 'text': 'Nice explanation!', 'timestamp': '2024-09-14T15:45:00'}], 'tags': ['MongoDB', 'Database', 'NoSQL', 'Tutorial']}


Query Comments

In [181]:
# Create an index on the "comments.user_id" field before querying for comments
posts.create_index([("comments.user_id", 1)])
print("Index created on 'comments.user_id' field!")


Index created on 'comments.user_id' field!


In [182]:
# Fetch all comments made by a specific user
user_id = "Bob"
comments_by_user = posts.find({"comments.user_id": user_id})
for post in comments_by_user:
    for comment in post["comments"]:
        if comment["user_id"] == user_id:
            print(comment)

{'user_id': 'Bob', 'text': 'Great post!', 'timestamp': '2024-09-12T10:00:00'}


#Part 4: Refactoring for Performance

In [183]:
from pymongo import MongoClient

# Connect to MongoDB
connection_string = "mongodb+srv://nabilnehme:nabilnehme@cluster0.51s8b.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0"
client = MongoClient(connection_string)
db = client['blog_platform']

# Define collections
users = db['users']
posts = db['posts']
comments = db['comments']

# Clear existing collections
users.delete_many({})
posts.delete_many({})
comments.delete_many({})
print("Collections cleared successfully!")

Collections cleared successfully!


In [184]:
# Insert 10 users
users_data = [
    {"_id": 1, "name": "Alice", "email": "alice@example.com"},
    {"_id": 2, "name": "Bob", "email": "bob@example.com"},
    {"_id": 3, "name": "Charlie", "email": "charlie@example.com"},
    {"_id": 4, "name": "David", "email": "david@example.com"},
    {"_id": 5, "name": "Emma", "email": "emma@example.com"},
    {"_id": 6, "name": "Frank", "email": "frank@example.com"},
    {"_id": 7, "name": "Grace", "email": "grace@example.com"},
    {"_id": 8, "name": "Hannah", "email": "hannah@example.com"},
    {"_id": 9, "name": "Isaac", "email": "isaac@example.com"},
    {"_id": 10, "name": "Jack", "email": "jack@example.com"}
]
users.insert_many(users_data)
print("10 users inserted successfully!")

10 users inserted successfully!


In [185]:
# Insert 20 unique posts
post_data = [
    {"_id": 101, "title": "Introduction to MongoDB", "content": "A beginner's guide to MongoDB databases.", "author_id": 1, "tags": ["MongoDB", "Database"]},
    {"_id": 102, "title": "Advanced Indexing in MongoDB", "content": "Understanding indexing strategies.", "author_id": 2, "tags": ["MongoDB", "Performance"]},
    {"_id": 103, "title": "Python and MongoDB", "content": "Using PyMongo to interact with MongoDB.", "author_id": 3, "tags": ["Python", "MongoDB"]},
    {"_id": 104, "title": "Data Modeling in NoSQL", "content": "Best practices for NoSQL schema design.", "author_id": 4, "tags": ["NoSQL", "Schema"]},
    {"_id": 105, "title": "Scaling MongoDB", "content": "Techniques for scaling large databases.", "author_id": 5, "tags": ["MongoDB", "Scaling"]},
    {"_id": 106, "title": "Aggregation Pipelines", "content": "How to use aggregation for analytics.", "author_id": 6, "tags": ["MongoDB", "Aggregation"]},
    {"_id": 107, "title": "Replicas and Sharding", "content": "High availability and horizontal scaling.", "author_id": 7, "tags": ["MongoDB", "Replication"]},
    {"_id": 108, "title": "Securing MongoDB", "content": "How to secure your MongoDB instance.", "author_id": 8, "tags": ["Security", "MongoDB"]},
    {"_id": 109, "title": "MongoDB Transactions", "content": "Using ACID transactions in MongoDB.", "author_id": 9, "tags": ["Transactions", "MongoDB"]},
    {"_id": 110, "title": "Optimizing Query Performance", "content": "How to optimize MongoDB queries.", "author_id": 10, "tags": ["Performance", "Query"]},
    {"_id": 111, "title": "Full-Text Search in MongoDB", "content": "Implementing full-text search.", "author_id": 1, "tags": ["Search", "MongoDB"]},
    {"_id": 112, "title": "MongoDB vs SQL Databases", "content": "Comparison between SQL and NoSQL.", "author_id": 2, "tags": ["Comparison", "Database"]},
    {"_id": 113, "title": "Backup and Restore Strategies", "content": "How to back up and restore MongoDB.", "author_id": 3, "tags": ["Backup", "MongoDB"]},
    {"_id": 114, "title": "Dockerizing MongoDB", "content": "Running MongoDB in Docker containers.", "author_id": 4, "tags": ["Docker", "MongoDB"]},
    {"_id": 115, "title": "Serverless MongoDB", "content": "Using MongoDB in a serverless architecture.", "author_id": 5, "tags": ["Serverless", "Cloud"]},
    {"_id": 116, "title": "Machine Learning with MongoDB", "content": "Integrating ML models with MongoDB.", "author_id": 6, "tags": ["Machine Learning", "MongoDB"]},
    {"_id": 117, "title": "Geospatial Queries", "content": "How to perform geospatial queries.", "author_id": 7, "tags": ["Geospatial", "MongoDB"]},
    {"_id": 118, "title": "Streaming Data with MongoDB", "content": "Using MongoDB for real-time streaming.", "author_id": 8, "tags": ["Streaming", "Real-Time"]},
    {"_id": 119, "title": "MongoDB Change Streams", "content": "Capturing data changes with Change Streams.", "author_id": 9, "tags": ["Change Streams", "MongoDB"]},
    {"_id": 120, "title": "Multi-Document Transactions", "content": "Ensuring consistency across documents.", "author_id": 10, "tags": ["Transactions", "ACID"]}
]
posts.insert_many(post_data)
print("20 posts inserted successfully!")

20 posts inserted successfully!


In [186]:
# Insert 50 unique comment per post (1000 comments)
comment_data = []
for post in post_data:
    post_id = post["_id"]
    for j in range(1, 51):  # 30 comments per post
        comment_data.append({
            "_id": (post_id * 100) + j,  # Ensure unique IDs
            "post_id": post_id,
            "user_id": ((post_id + j) % 10) + 1,  # Assign different users
            "text": f"User {((post_id + j) % 10) + 1} says: This is comment {j} on post '{post['title']}'.",
            "timestamp": f"2025-01-{(j % 30) + 1}T{10 + (j % 12)}:00:00"  # Cyclic timestamp variation
        })

comments.insert_many(comment_data)
print(f"{len(comment_data)} unique comments inserted successfully!")

1000 unique comments inserted successfully!


In [187]:
# Querying comments before indexing

import time

# Define the user ID to query
user_id = 1  # Example: Fetch comments made by user 1 (Alice)

### Step 1: Query comments **before indexing** and measure time
start_time = time.time()  # Start time before query
comments_by_user = list(comments.find({"user_id": user_id}))  # Convert cursor to list
end_time = time.time()  # End time after query

# Print time taken without index
print(f"Time taken WITHOUT index: {end_time - start_time:.6f} seconds")

# Print first few comments to verify
print("\nSample Comments (Without Index):")
for comment in comments_by_user[:5]:  # Print only the first 5 comments
    print(comment)

Time taken WITHOUT index: 0.356044 seconds

Sample Comments (Without Index):
{'_id': 10109, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 9 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-10T19:00:00'}
{'_id': 10119, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 19 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-20T17:00:00'}
{'_id': 10129, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 29 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-30T15:00:00'}
{'_id': 10139, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 39 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-10T13:00:00'}
{'_id': 10149, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 49 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-20T11:00:00'}


In [188]:
# Querying comments after indexing

### Step 2: Create an index on the `comments` collection for `user_id`
comments.create_index([("user_id", 1)])


### Step 3: Query comments **after indexing** and measure time
start_time = time.time()  # Start time before query
comments_by_user_indexed = list(comments.find({"user_id": user_id}))  # Convert cursor to list
end_time = time.time()  # End time after query

# Print time taken with index
print(f"Time taken WITH index: {end_time - start_time:.6f} seconds")

# Print first few comments to verify
print("\nSample Comments (With Index):")
for comment in comments_by_user_indexed[:5]:  # Print only the first 5 comments
    print(comment)


Time taken WITH index: 0.179362 seconds

Sample Comments (With Index):
{'_id': 10109, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 9 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-10T19:00:00'}
{'_id': 10119, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 19 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-20T17:00:00'}
{'_id': 10129, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 29 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-30T15:00:00'}
{'_id': 10139, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 39 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-10T13:00:00'}
{'_id': 10149, 'post_id': 101, 'user_id': 1, 'text': "User 1 says: This is comment 49 on post 'Introduction to MongoDB'.", 'timestamp': '2025-01-20T11:00:00'}


#Results

The results show a clear performance improvement with the use of an index:

  Time taken WITHOUT index: 0.358880 seconds

  Time taken WITH index: 0.179362 seconds


Indexing reduces the query time by about 50%. MongoDB indexes help optimize speed by enabling the database to find the relevant documents fast without scanning the full collection.

